### PR TITLE
Cache singularity containers for Galaxy tools

### PR DESCRIPTION
**Pull request by pcm32**
This introduces a mechanism to download all singularity containers for the tools in a Galaxy instance. It requires singularity and squashfs-tools for SIF image building accessible (in the server) to the user running the Galaxy server.

### CODE REVIEW

File: scripts/cache_mulled_singularity_containers.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    It will request the resolution of tools to containers from the instance,
+    and then for all mulled_singularity resolved tools that are pointing to
+    docker://<container-url>, it will download and build the singularity SIF.
+    This download is done only once per container, and applies after for all tools
+    that will make use of it.
+    """
+    args = get_args()
+    set_logging_level(debug=args.debug)
+    gi = GalaxyInstance(url=args.url, key=arg.key)
```
Comment (a-solovyev12) [inline scripts/cache_mulled_singularity_containers.py]: @pcm32 I'm not sure where the `arg.key` is coming from here? Is that supposed to be `args.api_key` ?

File: scripts/cache_mulled_singularity_containers.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    tools_deps = gi.make_get_request(gi.base_url + "/api/container_resolvers/toolbox").json()
+
+    # We keep a container url to tools, since the
+    # RAW API call to download a container requires a tool identifier (but only one).
+    container2tool_id = dict()
+    for tool_deps in tools_deps:
```
Comment (a-solovyev12) [inline scripts/cache_mulled_singularity_containers.py:48]: This naming makes it a little confusing to understand what the loop is iterating against. `for tool_dep in tool_deps:` would be easier to understand

Comment (a-solovyev12) [reply scripts/cache_mulled_singularity_containers.py:48]: Ah, never mind, I misread this, it looks correct to me.

File: scripts/cache_mulled_singularity_containers.py (hunk 1 above)
Comment (pcm32) [reply scripts/cache_mulled_singularity_containers.py]: Well spotted! I was doing some changes that I shouldn't have done after testing.
```suggestion
    gi = GalaxyInstance(url=args.url, key=args.key)
```

File: scripts/cache_mulled_singularity_containers.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+    It will request the resolution of tools to containers from the instance,
+    and then for all mulled_singularity resolved tools that are pointing to
+    docker://<container-url>, it will download and build the singularity SIF.
+    This download is done only once per container, and applies after for all tools
+    that will make use of it.
+    """
+    args = get_args()
+    set_logging_level(debug=args.debug)
+    gi = GalaxyInstance(url=args.url, key=args.key)
```
Comment (a-solovyev12) [inline scripts/cache_mulled_singularity_containers.py]: To match it with the argument: 
```suggestion
    gi = GalaxyInstance(url=args.url, key=args.api_key)
```

File: scripts/cache_mulled_singularity_containers.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+def set_logging_level(debug=False):
+    logging.basicConfig(
+        level=logging.DEBUG if debug else logging.INFO,
+        format='%(asctime)s - %(message)s',
+        datefmt='%d-%m-%y %H:%M:%S')
+
+def main():
+    """
+    Downloads singularity containers on a Galaxy instance which is configured
+    to use such dependency resolvers (mulles_singularity and cached_mulled_singularity).
```
Comment (pcm32) [inline scripts/cache_mulled_singularity_containers.py]:
```suggestion
    to use such dependency resolvers (mulled_singularity and cached_mulled_singularity).
```

**Comment by pcm32** [comment]
Thanks @a-solovyev12! Will re-test this and then merge.

**Review by a-solovyev12 — APPROVED** [review]

**Comment by pcm32** [comment]
Re-tested with latest changes.

**Pull request merged**